File: Leetcode/Leetcode-DSA-course/Linked-List/24-swap-nodes-in-pairs/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if (not head) or (not head.next):
            return head
        prev = None
        dummy = head.next
        while head and head.next:
            if prev:
                prev.next = head.next
            prev = head
            nextNode = head.next.next
            head.next.next = head
            head.next = nextNode
            head = nextNode
        return dummy

File: Leetcode/Leetcode-DSA-course/Linked-List/24-swap-nodes-in-pairs/swap-nodes-in-pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null)
            return head;

        ListNode prev = null;
        ListNode dummy = head.next;
        while (head != null && head.next != null) {
            if (prev != null)
                prev.next = head.next;
            prev = head;
            ListNode nextNode = head.next.next;
            head.next.next = head;
            head.next = nextNode;
            head = nextNode;
        }
        return dummy;
    }
}